share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java

```diff
@@ -17,58 +17,536 @@
+        Optional<CoordinatorResult<WriteShareGroupStateResponseData, CoordinatorRecord>> error = maybeGetWriteStateError(request);
+        if (error.isPresent()) {
+            return error.get();
+        }
+
+        String groupId = request.groupId();
+        WriteShareGroupStateRequestData.WriteStateData topicData = request.topics().get(0);
+        WriteShareGroupStateRequestData.PartitionData partitionData = topicData.partitions().get(0);
+
+        SharePartitionKey key = SharePartitionKey.getInstance(groupId, topicData.topicId(), partitionData.partition());
+        List<CoordinatorRecord> recordList;
+
+        if (!shareStateMap.containsKey(key)) {
+            // since this is the first time we are getting a write request, we should be creating a share snapshot record
+            recordList = Collections.singletonList(ShareCoordinatorRecordHelpers.newShareSnapshotRecord(
+                groupId, topicData.topicId(), partitionData.partition(), ShareGroupOffset.fromRequest(partitionData)
+            ));
+        } else if (snapshotUpdateCount.getOrDefault(key, 0) >= snapshotUpdateRecordsPerSnapshot) {
+            // Since the number of update records for this share part key exceeds snapshotUpdateRecordsPerSnapshot,
+            // we should be creating a share snapshot record.
+            List<PersisterOffsetsStateBatch> batchesToAdd;
+            if (partitionData.startOffset() == -1) {
+                batchesToAdd = combineStateBatches(
+                    shareStateMap.get(key).stateBatchAsSet(),
+                    partitionData.stateBatches().stream()
+                        .map(PersisterOffsetsStateBatch::from)
+                        .collect(Collectors.toCollection(LinkedHashSet::new)));
+            } else {
+                // start offset is being updated - we should only
+                // consider new updates to batches
+                batchesToAdd = partitionData.stateBatches().stream()
+                    .map(PersisterOffsetsStateBatch::from).collect(Collectors.toList());
+            }
+
+            int newLeaderEpoch = partitionData.leaderEpoch() == -1 ? shareStateMap.get(key).leaderEpoch() : partitionData.leaderEpoch();
+            int newStateEpoch = partitionData.stateEpoch() == -1 ? shareStateMap.get(key).stateEpoch() : partitionData.stateEpoch();
+            long newStartOffset = partitionData.startOffset() == -1 ? shareStateMap.get(key).startOffset() : partitionData.startOffset();
+
+            recordList = Collections.singletonList(ShareCoordinatorRecordHelpers.newShareSnapshotRecord(
+                groupId, topicData.topicId(), partitionData.partition(),
+                new ShareGroupOffset.Builder()
+                    .setStartOffset(newStartOffset)
+                    .setLeaderEpoch(newLeaderEpoch)
+                    .setStateEpoch(newStateEpoch)
+                    .setStateBatches(batchesToAdd)
+                    .build()));
+            snapshotUpdateCount.put(key, 0);
```

Comment:
We should not update a timeline structure here. Recall that this method generate the records and eventual response for the proposed write. We cannot update our in-memory state until the write is committed. 

It looks like `snapshotUpdateCount` is used to determine when we should write a snapshot instead of a delta (L299). We can reset this counter to zero in handleShareSnapshot as we're replaying records